networkx/algorithms/approximation/steinertree.py

```diff
@@ -248,3 +249,179 @@ def steiner_tree(G, terminal_nodes, weight="weight", method=None):
         )
     T = G.edge_subgraph(edges)
     return T
+
+
+def _reachable_within(G, source, cutoff):
+    """Return the set of nodes reachable from source within cutoff depth."""
+    seen = {source}
+    q = deque([(source, 0)])
+    while q:
+        node, depth = q.popleft()
+        if depth == cutoff:
+            continue
+        for nbr in G[node]:
+            if nbr not in seen:
+                seen.add(nbr)
+                q.append((nbr, depth + 1))
+    return seen
+
+
+def _directed_steiner_tree_density(G, terminals, weight):
+    """Compute density = total cost / #covered terminals."""
+    if G.number_of_edges() == 0:
+        return float("inf")
+
+    terminals = set(terminals)
+    num_terminals = 0
+    total = 0
+    for u, v, d in G.edges(data=True):
+        if u in terminals:
+            terminals.remove(u)
+            num_terminals += 1
+        if v in terminals:
+            terminals.remove(v)
+            num_terminals += 1
+        total += d.get(weight, 1)
+
+    return float("inf") if num_terminals == 0 else total / num_terminals
+
+
+def _directed_steiner_tree(G, root, terminals, min_terminals, cutoff, weight):
+    """Recursive helper for directed_steiner_tree."""
+    H = nx.DiGraph()
+    if min_terminals <= 0 or not terminals or cutoff == 0:
+        return H
```

Comment:
This is a private functuion so we shouldn't need to check the inputs. They've already been checked before calling this, right?